Mu2eKinKal/src/LoopHelixFit_module.cc

```diff
@@ -347,88 +488,103 @@ namespace mu2e {
+            if(!ktrk) {
+              ktrk = std::move(ktrk_i);
+              helix_dir = dir;
+            } else {
+              const bool accept = compare_tracks(ktrk, ktrk_i) == 0;
+              // Print out verbose info if requested
+              if(print_ > 0)
+                printf("[LoopHelixFit::%s::%s]: Fit both direction options, slope = %8.1e, sig = %5.2f, status current = %i, status new = %i, p(chi^2) current = %.4f, p(chi^2) new = %.4f: accept new track = %o\n",
+                       __func__, moduleDescription().moduleLabel().c_str(), hseed.recoDir().slope(), hseed.recoDir().slopeSig(),
+                       ktrk && goodFit(*ktrk), ktrk_i && goodFit(*ktrk_i),
+                       (ktrk  ) ? ktrk  ->fitStatus().chisq_.probability() : -1.,
+                       (ktrk_i) ? ktrk_i->fitStatus().chisq_.probability() : -1., accept);
+              if(accept) {
+                ktrk = std::move(ktrk_i);
+                helix_dir = dir; // store the propagation direction
               }
             }
-            // extrapolate as required
-            if(goodfit && extrapolate_) extrapolate(*ktrk);
-            if(print_>0)ktrk->printFit(std::cout,print_);
-            if(goodfit || saveall_){
-              TrkFitFlag fitflag(hptr->status());
-              fitflag.merge(fitflag_);
-              if(goodfit)
-                fitflag.merge(TrkFitFlag::FitOK);
-              else
-                fitflag.clear(TrkFitFlag::FitOK);
-              auto kkseed = kkfit_.createSeed(*ktrk,fitflag,*calo_h);
-              sampleFit(*ktrk,kkseed._inters);
-              kkseedcol->push_back(kkseed);
-              // fill assns with the helix seed
-              auto hptr = art::Ptr<HelixSeed>(hseedcol_h,iseed);
-              auto kseedptr = art::Ptr<KalSeed>(KalSeedCollectionPID,kkseedcol->size()-1,KalSeedCollectionGetter);
-              kkseedassns->addSingle(kseedptr,hptr);
-              // save (unpersistable) KKTrk in the event
-              ktrkcol->push_back(ktrk.release());
+          }
+        }
+
+        if(!ktrk) continue; //ensure that the track exists
+
+        // Check the fit
+        auto goodfit = goodFit(*ktrk);
+
+        // extrapolate as required
+        if(goodfit && extrapolate_) extrapolate(*ktrk);
+        if(print_>1) ktrk->printFit(std::cout,print_-1);
+
+        // save the fit result
+        if(goodfit || saveall_){
+          auto hptr = HPtr(hseedcol_h,iseed);
+          TrkFitFlag fitflag(hptr->status());
+          fitflag.merge(fitflag_);
+          if(goodfit)
+            fitflag.merge(TrkFitFlag::FitOK);
+          else
+            fitflag.clear(TrkFitFlag::FitOK);
+          if(undefined_dir) fitflag.merge(TrkFitFlag::AmbFitDir);
+          auto kkseed = kkfit_.createSeed(*ktrk,fitflag,*calo_h);
+          if(print_>0) printf("[LoopHelixFit::%s::%s] Create seed             : fitcon = %.4f, nHits = %2lu, seedActiveHits = %2u, %lu calo-hits\n",
```

Comment:
these additional print lines make it hard to read the code, for instance finding the sampleFit() call. Please move the prints into a single function.

Reply:
I agree, I was adding these mostly for the debugging process to understand if I was change any track information by fitting once vs. twice. I've moved these to a `print_track_info` function, which I can remove or cleanup more now that the debugging has settled.